bee-network/bee-gossip/src/handshake.rs

```diff
@@ -0,0 +1,380 @@
+            }
+
+            if num_received > MAX_HANDSHAKE_PACKET_SIZE {
+                return Err(HandshakeError::PacketSizeMismatch {
+                    received: num_received,
+                    max_allowed: MAX_HANDSHAKE_PACKET_SIZE,
+                });
+            }
+
+            let packet = Packet::from_protobuf(&buf[..num_received]).map_err(HandshakeError::Decode)?;
+            let packet_type = packet.ty().map_err(HandshakeError::PacketType)?;
+
+            if matches!(packet_type, PacketType::Handshake) {
+                log::info!("received handshake response.");
+
+                let res = HandshakeResponse::from_protobuf(packet.data())?;
+
+                let peer_res_data = res.protobuf()?;
+
+                HandshakeValidator::validate_response(&peer_res_data, &local_req_data)?;
+
+                log::debug!("handshake response is valid");
+
+                break packet;
+            }
+        }
+    };
+
+    let peer_public_key = packet.public_key();
+    let mut pk = [0u8; 32];
+    pk.copy_from_slice(&peer_public_key[..32]);
+    let peer_public_key = ed25519::PublicKey::try_from_bytes(pk).map_err(HandshakeError::PublicKey)?;
+    let peer_identity = PeerId::from_public_key(peer_public_key);
+
+    Ok(peer_identity)
+}
+
+/// Errors, that may occur during the handshake process.
+#[derive(Debug)]
+pub enum HandshakeError {
+    /// The peer dropped the connection.
+    ConnectionResetByPeer,
+    /// An I/O error occurred during handshaking.
+    Io(io::Error),
+    /// A protobuf encode error occurred.
+    Encode(prost::EncodeError),
+    /// A protobuf decode error occurred.
+    Decode(prost::DecodeError),
+    /// The peer doesn't use the same protocol version.
+    VersionMismatch {
+        /// The expected version.
+        expected: u32,
+        /// The received version.
+        received: u32,
+    },
+    /// The handshake timestamp was too old.
+    Expired,
+    /// The peer didn't provide the expected hash in its handshake response.
+    RequestHashMismatch,
+    /// The handshake response came in too late.
+    _ResponseTimeout,
+    /// Wrong packet type.
+    PacketType(io::Error),
+    /// The peer send a packet which is too large.
+    PacketSizeMismatch {
+        /// The received packet size.
+        received: usize,
+        /// The maximum allowed packet size.
+        max_allowed: usize,
+    },
+    /// TODO
+    PublicKey(crypto::Error),
+}
+
+/// Represents a handshake request, and is a wrapper of the corresponding Protobuf type.
+pub struct HandshakeRequest {
+    inner: proto::HandshakeRequest,
+}
+
+impl HandshakeRequest {
+    /// Creates a new handshake request.
+    pub fn new(to: IpAddr) -> Self {
+        let inner = proto::HandshakeRequest {
+            version: VERSION,
+            to: to.to_string(),
+            timestamp: SystemTime::now()
+                .duration_since(UNIX_EPOCH)
+                .expect("system clock error")
```

Comment:
I wonder if we should normalize ` timestamp: SystemTime::now().duration_since(UNIX_EPOCH).expect("here be dragons")` in a crate like `bee-common` or `bee-utils`.

Reply:
sounds good to me :+1: